### PR TITLE
Migrate MSL hints to main SNOS repository

### PR DESCRIPTION
**Pull request by odesenfans**
## Type

- [x] feature
- [ ] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Description

Migration of the hints already implemented by Moonsong Labs to the main SNOS repository. We will use the main repository moving forward.

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: src/hints/state.rs [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    vm.insert_value(bottom_addr, node_values[2])?;
+
+    // TODO: prevent underflow (original hint doesn't appear to care)?
+    // compute `ids.hash_ptr.result = ids.node - ids.edge.length`
+    let res = node - node_values[0];
+
+    // ids.hash_ptr refers to SpongeHashBuiltin (see cairo-lang's sponge_as_hash.cairo)
+    // it is a struct with 6 felts and `result` is the 4th
+    let hash_ptr = get_relocatable_from_var_name(vars::ids::HASH_PTR, vm, ids_data, ap_tracking)?;
+    let hash_ptr_result: Relocatable = (hash_ptr + 3)?;
```
Comment (maciejka) [inline src/hints/state.rs:194]: Seems fragile.

File: src/hints/state.rs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+            (vars::ids::FINAL_ROOT.to_string(), HintReference::new_simple(-2)),
+            (vars::ids::MERKLE_HEIGHT.to_string(), HintReference::new_simple(-1)),
+        ]);
+        insert_value_from_var_name(vars::ids::MERKLE_HEIGHT, 251_usize, &mut vm, &ids_data, &ap_tracking)
+            .expect("Couldn't insert 252 into ids.MERKLE_HEIGHT");
+
+        let mut exec_scopes: ExecutionScopes = Default::default();
+        exec_scopes.insert_value(vars::scopes::OS_INPUT, os_input);
+
+        set_preimage_for_state_commitments(&mut vm, &mut exec_scopes, &ids_data, &ap_tracking, &constants).unwrap();
```
Comment (maciejka) [inline src/hints/state.rs:257]: Shouldn't we call a hint through hint processor to verify it is properly registered there?

File: src/hints/state.rs (hunk 1 above)
Comment (odesenfans) [reply src/hints/state.rs]: This should be fixed now, I merged the offset PR.

File: src/hints/state.rs (hunk 2 above)
Comment (odesenfans) [reply src/hints/state.rs:253]: This is a unit test so I don't really care about that here. If a hint is not registered integration tests will not work and the error is easy to spot and resolve, so I'd rather keep the unit tests focused.

**Comment by odesenfans** [comment]
@maciejka Can you approve? I renamed the hint.

**Review by maciejka — APPROVED** [review]

**Pull request merged**